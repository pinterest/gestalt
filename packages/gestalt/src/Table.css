.table {
  border-collapse: collapse;
  border-spacing: 0;
  width: 100%;
}

.th {
  composes: paddingY3 from "./boxWhitespace.css";
  composes: paddingX3 from "./boxWhitespace.css";
}

.td {
  composes: paddingY3 from "./boxWhitespace.css";
  composes: paddingX3 from "./boxWhitespace.css";
}

.thead {
  composes: borderBottom from "./Borders.css";
}
