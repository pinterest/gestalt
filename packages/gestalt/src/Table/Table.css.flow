// @flow strict
/* This file is automatically generated by css-modules-flow-types */
declare module.exports: {|
  +'table': string,
  +'td': string,
  +'th': string,
  +'thead': string,
|};
