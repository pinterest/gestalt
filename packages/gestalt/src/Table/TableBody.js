// @flow strict
import * as React from 'react';

type Props = {|
  children: React.Node,
|};

export default function TableBody(props: Props) {
  return <tbody>{props.children}</tbody>;
}
