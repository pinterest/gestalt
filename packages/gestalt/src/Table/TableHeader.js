// @flow strict
import * as React from 'react';
import styles from './Table.css';

type Props = {|
  children: React.Node,
|};

export default function TableHeader(props: Props) {
  return <thead className={styles.thead}>{props.children}</thead>;
}
