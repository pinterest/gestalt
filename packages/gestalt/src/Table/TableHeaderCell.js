// @flow strict
import * as React from 'react';
import styles from './Table.css';

type Props = {|
  children: React.Node,
  colSpan?: number,
  rowSpan?: number,
  scope?: 'col' | 'colgroup' | 'row' | 'rowgroup',
|};

export default function TableHeaderCell(props: Props) {
  const { children, colSpan, scope, rowSpan } = props;

  return (
    <th
      className={styles.th}
      scope={scope || 'col'}
      colSpan={colSpan}
      rowSpan={rowSpan}
    >
      {children}
    </th>
  );
}
