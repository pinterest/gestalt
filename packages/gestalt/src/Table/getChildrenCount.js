// @flow strict
import { Children, Fragment, type Node } from 'react';

function getChildrenCount(children: Node): number {
  // $FlowFixMe[underconstrained-implicit-instantiation]
  const topChildren = Children.toArray(children);
  let nestedChildrenCount = 0;

  topChildren.forEach((child) => {
    // We need to check for Fragment first, so we can check for nested Table.Cell that don't get counted in Children.toArray
    if (child?.type === Fragment) {
      const nestedChildren = child.props.children.reduce(
        (accumulator, currentValue) =>
          currentValue?.type?.displayName === 'Table.Cell' ? 1 + accumulator : accumulator,
        -1,
      );
      nestedChildrenCount += nestedChildren;
    }
  });

  return topChildren.length + nestedChildrenCount;
}

export default getChildrenCount;
