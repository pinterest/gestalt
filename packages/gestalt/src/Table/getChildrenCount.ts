import { Children, Fragment, ReactNode } from 'react';

function getChildrenCount(children: ReactNode): number {
  const topChildren = Children.toArray(children);
  let nestedChildrenCount = 0;

  topChildren.forEach((child) => {
    // We need to check for Fragment first, so we can check for nested Table.Cell that don't get counted in Children.toArray
    // @ts-expect-error - TS2339 - Property 'type' does not exist on type 'string | number | ReactElement<any, string | JSXElementConstructor<any>> | Iterable<ReactNode> | ReactPortal'.
    if (child?.type === Fragment) {
      // @ts-expect-error - TS2339 - Property 'props' does not exist on type 'string | number | ReactElement<any, string | JSXElementConstructor<any>> | Iterable<ReactNode> | ReactPortal'.
      const nestedChildren = child.props.children.reduce(
        // @ts-expect-error - TS7006 - Parameter 'accumulator' implicitly has an 'any' type. | TS7006 - Parameter 'currentValue' implicitly has an 'any' type.
        (accumulator, currentValue) =>
          currentValue?.type?.displayName === 'Table.Cell' ? 1 + accumulator : accumulator,
        -1,
      );
      nestedChildrenCount += nestedChildren;
    }
  });

  return topChildren.length + nestedChildrenCount;
}

export default getChildrenCount;
