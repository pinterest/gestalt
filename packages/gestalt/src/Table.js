// @flow strict
import * as React from 'react';
import Box from './Box.js';
import styles from './Table.css';
import TableCell from './TableCell.js';
import TableBody from './TableBody.js';
import TableHeader from './TableHeader.js';
import TableHeaderCell from './TableHeaderCell.js';
import TableRow from './TableRow.js';
import TableFooter from './TableFooter.js';

type Props = {|
  children: React.Node,
  borderSize?: 'sm' | 'none',
|};

export default function Table(props: Props) {
  const { borderSize, children } = props;

  return (
    <Box
      overflow="scrollX"
      {...(borderSize === 'sm' ? { borderSize: 'sm', rounding: 1 } : {})}
    >
      <table className={styles.table}>{children}</table>
    </Box>
  );
}

Table.Body = TableBody;

Table.Cell = TableCell;

Table.Footer = TableFooter;

Table.Header = TableHeader;

Table.HeaderCell = TableHeaderCell;

Table.Row = TableRow;
