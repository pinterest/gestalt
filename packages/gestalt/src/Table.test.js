// @flow strict
import React from 'react';
import renderer from 'react-test-renderer';
import Table from './Table.js';

test('renders correctly', () => {
  const tree = renderer
    .create(
      <Table>
        <div>rest of table</div>
      </Table>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly with border', () => {
  const tree = renderer
    .create(
      <Table borderSize="sm">
        <div>rest of table</div>
      </Table>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
