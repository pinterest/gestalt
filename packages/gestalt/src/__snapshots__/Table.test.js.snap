// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="box overflowScrollX"
>
  <table
    className="table"
  >
    <div>
      rest of table
    </div>
  </table>
</div>
`;

exports[`renders correctly with border 1`] = `
<div
  className="borderColorLightGray box overflowScrollX rounding1 sizeSm solid"
>
  <table
    className="table"
  >
    <div>
      rest of table
    </div>
  </table>
</div>
`;
